### PR TITLE
fix assisted decoding assistant model inputs

### PR DESCRIPTION
**Pull request by jiqing-feng**
In the last [PR](https://github.com/huggingface/transformers/pull/26892), we didn't consider the `decoder_attention_mask` while updating `model_kwargs`, see [here](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L4742-L4747). This PR has fixed it.

Furthermore, I also use a cleaner way to process assistant models's inputs.

Hi @gante  , would you please help me to review this PR? Thx!

### CODE REVIEW

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:4562]:
```suggestion
```
(Note: `breakpoint()` is easier to write and achieves the same :D )

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:4575]:
```suggestion
                has_past_key_values = assistant_inputs.get("past_key_values", None) is not None
                if has_past_key_values and assistant_inputs[input_ids_key].shape[-1] not in (1, 2):
```
(We prefer avoiding acronyms unless they are prohibitively long. It helps with readability :) )

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:4587]: This "1.5" should start in `if assistant_kwargs ...`. The two lines below are still part of "greedily select the next candidate token"

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:4606]: We can remove this line if we change 

```
            model_kwargs["assistant_past_key_values"] = _crop_past_key_values(
                assistant_model, model_kwargs["assistant_past_key_values"], new_cache_size - 1
            )  # the assistant does not have the token after the last match, hence the -1
```

a few lines below into

```
            assistant_kwargs["assistant_past_key_values"] = _crop_past_key_values(
                assistant_model, assistant_kwargs["assistant_past_key_values"], new_cache_size - 1
            )  # the assistant does not have the token after the last match, hence the -1
```

That way, we would have no assistant kwargs in `model_kwargs`, which would be great!

**Review by gante — COMMENTED** [review]
@jiqing-feng thank you for promptly working on a fix! 🤗 

I suspected this could be an issue, but I was a bit lazy and decided to rely on the tests. The catch is that the tests are stochastic (they are sample-based), so we has a lucky run in the past CI.

Regarding the PR itself: a few minor nits, and then it should be ready to go!

**Comment by gante** [comment]
@jiqing-feng If possible, I would also like to revert [these temporary changes](https://github.com/huggingface/transformers/pull/27511) in this PR :)

**Comment by ArthurZucker** [comment]
🤗 thanks for the fix we had to skip it in #27508 as well! (Only the relevant test)

File: src/transformers/generation/utils.py
Comment (jiqing-feng) [reply src/transformers/generation/utils.py:4606]: It is not possible. `model_kwargs` should always have `assistant_past_key_values` because we copied `assistant_kwargs` from `model_kwargs` in each round.

Comment (jiqing-feng) [reply src/transformers/generation/utils.py:4606]: Oh, I see your point. There is no need to copy `model_kwargs` in every round.

**Comment by jiqing-feng** [comment]
Hi @gante @ArthurZucker . I think I have fixed all the comments and also added the tests you mentioned. Would you please help me review it? Thx!

BTW, the failed test if not related to my changes.

**Review by gante — APPROVED** [review]
Thank you for iterating 👍

File: tests/models/speech_to_text/test_modeling_speech_to_text.py
Comment (gante) [inline tests/models/speech_to_text/test_modeling_speech_to_text.py:762]: Actually, we need to keep this skip, it is what is causing the failure in CI!

Comment (jiqing-feng) [reply tests/models/speech_to_text/test_modeling_speech_to_text.py]: Done.

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! 

Before we merge can we update the non-slow test to be more robust? If the past CI was green because of a lucky sample, how do we know that this PR fixes it and this wasn't just another lucky run? e.g. can we set a seed which we know causes a failure on main and passes here?

**Comment by gante** [comment]
@amyeroberts I'll have to think harder about assisted generation test robustness, as there are two conflicting effects in place:
1. In theory, assisted generation should yield the exact same outputs
2. In practice, due to the matrix multiplication being shape-dependent (see [here](https://github.com/huggingface/transformers/issues/25420#issuecomment-1775317535)), there will be tiny fluctuations. With random models, this means that the odds of a simple assisted vs non-assisted output check failing are high.

On top of that, pinning a seed to a previous failure does not prevent failure check in new models or flags. 

My suggestion would be: I'll work on test robustness today, and we merge this fix as is. WDYT?

**Comment by amyeroberts** [comment]
> In practice, due to the matrix multiplication being shape-dependent (see https://github.com/huggingface/transformers/issues/25420#issuecomment-1775317535), there will be tiny fluctuations. With random models, this means that the odds of a simple assisted vs non-assisted output check failing are high.

For my own understanding, why wouldn't a seed resolve the issues in randomness here? I'm guessing the tests are using `hf-internal-testing/tiny-random-model-name` which can change? 

> On top of that, pinning a seed to a previous failure does not prevent bad failure checks in future models or flags.

Agreed - but it should make sure that this one passes! For any future models or flags we should add new tests. 

In terms of tests to add - this relates back to my previous [request here](https://github.com/huggingface/transformers/pull/26892#issuecomment-1804158551). It seems that the PR broke for a specific model type (encoder-decoder). Are there tests, which do not rely on randomness, which we can add that make sure just the API works?

**Comment by patrickvonplaten** [comment]
Hey @jiqing-feng,

There is sadly still a bug with speculative decoding. The following doesn't work:

```py
from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor, pipeline
from transformers import AutoModelForCausalLM
from datasets import load_dataset
import time
import torch

device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

model_id = "openai/whisper-large-v2"

model = AutoModelForSpeechSeq2Seq.from_pretrained(
    model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
model.to(device)

processor = AutoProcessor.from_pretrained(model_id)

assistant_model_id = "distil-whisper/distil-large-v2"
assistant_model = AutoModelForCausalLM.from_pretrained(
    assistant_model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
assistant_model.to(device)

dataset = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
sample = dataset[0]["audio"]

input_features = processor(sample["array"], return_tensors="pt").input_features.to("cuda").to(torch.float16)

# warm-up
_ = model.generate(input_features, assistant_model=assistant_model)

start_time = time.time()
out = model.generate(input_features, assistant_model=assistant_model)
# out = model.generate(input_features)
print(time.time() - start_time)
```

**Comment by jiqing-feng** [comment]
Hi @patrickvonplaten 

I run the following script on my CPU device, and it works well
```python
from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor, pipeline
from transformers import AutoModelForCausalLM
from datasets import load_dataset
import time
import torch

device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

model_id = "openai/whisper-large-v2"

model = AutoModelForSpeechSeq2Seq.from_pretrained(
    model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
model.to(device)

processor = AutoProcessor.from_pretrained(model_id)

assistant_model_id = "openai/whisper-tiny"
assistant_model = AutoModelForSpeechSeq2Seq.from_pretrained(
   assistant_model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
assistant_model.to(device)

dataset = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
sample = dataset[0]["audio"]

input_features = processor(sample["array"], return_tensors="pt").input_features.to("cuda").to(torch.float16)

# warm-up
_ = model.generate(input_features, assistant_model=assistant_model)

start_time = time.time()
out = model.generate(input_features, assistant_model=assistant_model)
# out = model.generate(input_features)
print(time.time() - start_time)
```

**Comment by jiqing-feng** [comment]
Hi @patrickvonplaten 

I run the following script on my CPU device, and it works well.
```python
from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor, pipeline
from transformers import AutoModelForCausalLM
from datasets import load_dataset
import time
import torch

device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

model_id = "openai/whisper-large-v2"

model = AutoModelForSpeechSeq2Seq.from_pretrained(
    model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
model.to(device)

processor = AutoProcessor.from_pretrained(model_id)

assistant_model_id = "openai/whisper-tiny"
assistant_model = AutoModelForSpeechSeq2Seq.from_pretrained(
   assistant_model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
assistant_model.to(device)

dataset = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
sample = dataset[0]["audio"]

input_features = processor(sample["array"], return_tensors="pt").input_features.to(device).to(torch_dtype)

# warm-up
_ = model.generate(input_features, assistant_model=assistant_model)

start_time = time.time()
out = model.generate(input_features, assistant_model=assistant_model)
# out = model.generate(input_features)
print(time.time() - start_time)
```

**Comment by patrickvonplaten** [comment]
Hey @jiqing-feng,

Thanks so much for quickly jumping on fixing the problem here :pray: 

It sadly still doesn't fix Whisper distillation as per code snippet above. To make sure distil whisper works again on "main", we have now reverted the PR here: https://github.com/huggingface/transformers/pull/27523 and also added two slow tests that should be run now everytime we do changes to assisted decoding:

```py
RUN_SLOW=1 pytest tests/models/whisper/test_modeling_whisper.py -k "distil" -sv
```

It would be amazing if you could maybe try to open a new PR that is rebased to current "main" with all your nice changes and in which all fast tests as well as the slow tests pass:

```py
RUN_SLOW=1 pytest tests/models/whisper/test_modeling_whisper.py -k "distil" -sv
```

**Comment by jiqing-feng** [comment]
Hi @patrickvonplaten . There is no need to open a new PR. I have fixed the conflicts. 

There might be a mistake ([here](https://github.com/huggingface/transformers/blob/main/tests/models/whisper/test_modeling_whisper.py#L1752-L1755)) that I see that you use `distil-whisper/distil-large-v2` as assistant model, and you use `WhisperForCausalLM` to load a `WhisperForConditionalGeneration` model. The `distil-whisper/distil-large-v2` is an encoder-decoder model, so I use `WhisperForConditionalGeneration` to load it. After this change, I can successfully run **RUN_SLOW=1 pytest tests/models/whisper/test_modeling_whisper.py -k "distil" -sv** on my current changes.

**Comment by gante** [comment]
@jiqing-feng the improved assisted generation tests were merged 🤗

**Comment by jiqing-feng** [comment]
Hi @gante . I also updated my code base. Would you please help to merge this PR? Thx.

**Comment by gante** [comment]
Hi @jiqing-feng 👋 

I got it working on my end, without the change you added on the Whisper test (which we must revert). It is a non-trivial set of changes, so I'm going to detail the entire diff :)

1. Remove the `self._extend_attention_mask` and `self._extend_token_type_ids` functions from the `GenerationMixin`
2. Replace them by the following stand-alone functions, which can be added at the bottom of the file
```py
def _prepare_attention_mask(model_kwargs: Dict[str, Any], new_length: int, is_encoder_decoder: bool) -> Dict[str, Any]:
    """Expands or crops the model's mask for decoding purposes, to the defined length"""

    mask_key = "decoder_attention_mask" if is_encoder_decoder else "attention_mask"
    if mask_key not in model_kwargs:
        return model_kwargs

    mask = model_kwargs[mask_key]
    mask_length_diff = new_length - mask.shape[1]

    if mask_length_diff < 0:
        model_kwargs[mask_key] = mask[:, :mask_length_diff]
    elif mask_length_diff > 0:
        model_kwargs[mask_key] = torch.cat([mask, mask.new_ones((mask.shape[0], mask_length_diff))], dim=-1)
    return model_kwargs


def _prepare_token_type_ids(model_kwargs: Dict[str, Any], new_length: int) -> Dict[str, Any]:
    """Expands or crops the model's token_type_ids for decoding purposes, to the defined length"""
    if "token_type_ids" not in model_kwargs or model_kwargs["token_type_ids"] is None:
        return model_kwargs

    token_type_ids = model_kwargs["token_type_ids"]
    final_token_type = token_type_ids[:, -1].unsqueeze(-1)
    type_length_diff = new_length - token_type_ids.shape[1]

    if type_length_diff < 0:
        token_type_ids = token_type_ids[:, :type_length_diff]
    elif type_length_diff > 0:
        token_type_copies = final_token_type.repeat(1, type_length_diff)
        model_kwargs["token_type_ids"] = torch.cat([model_kwargs["token_type_ids"], token_type_copies], dim=-1)
    return model_kwargs
```

3. Replace the code after `# Update assistant_kwargs for the assistant's next round of generations` by
```py
            assistant_kwargs = _prepare_attention_mask(
                assistant_kwargs, new_cur_len, assistant_model.config.is_encoder_decoder
            )
            assistant_kwargs = _prepare_token_type_ids(assistant_kwargs, new_cur_len)
```

4. Replace the code after `# 2.1. Prepare the model inputs` by
```py
            candidate_kwargs = copy.copy(model_kwargs)
            candidate_kwargs = _prepare_attention_mask(
                candidate_kwargs, candidate_input_ids.shape[1], self.config.is_encoder_decoder
            )
            candidate_kwargs = _prepare_token_type_ids(candidate_kwargs, candidate_input_ids.shape[1])

            model_inputs = self.prepare_inputs_for_generation(candidate_input_ids, **candidate_kwargs)
```

5. Replace the code after `# prepare assistant model's keys of inputs` by
```py
        assistant_kwargs = copy.copy(model_kwargs)
        if assistant_model.config.is_encoder_decoder:
            # both are encoder-decoder
            input_ids_key = "decoder_input_ids"
            attention_key = "decoder_attention_mask"
            assistant_kwargs["encoder_outputs"] = assistant_kwargs.pop("assistant_encoder_outputs")
        elif "assistant_encoder_outputs" in assistant_kwargs:
            # special case for encoder-decoder with decoder-only assistant (like DistilWhisper)
            input_ids_key = "input_ids"
            attention_key = "attention_mask"
            assistant_kwargs["attention_mask"] = assistant_kwargs.get(
                "decoder_attention_mask",
                torch.ones((input_ids.shape[0], 1), device=input_ids.device, dtype=torch.long),
            )
            assistant_kwargs["encoder_outputs"] = assistant_kwargs.pop("assistant_encoder_outputs")
        else:
            # both are decoder-only
            input_ids_key = "input_ids"
            attention_key = "attention_mask"
```

All these changes will make `assisted_generation` compatible with all use cases, even the more complex DistilWhisper 🤗

**Comment by jiqing-feng** [comment]
Hi @gante . Thanks for your review, I have updated all the changes you proposed. Would you please help me to check and merge it? Thx!

File: tests/models/whisper/test_modeling_whisper.py
Comment (patrickvonplaten) [inline tests/models/whisper/test_modeling_whisper.py:1751]: The test should be kept as is here, we only want to load the decoder of distil-large-v2 here

Comment (jiqing-feng) [reply tests/models/whisper/test_modeling_whisper.py]: OK, I have reverted it.

**Review by gante — APPROVED** [review]
Perfect, thank you for working on the changes 💪 

@jiqing-feng if possible, it would be nice to delete the now unused `_extend_attention_mask` and `_extend_token_type_ids` functions :)

@amyeroberts I've confirmed on my end that all relevant tests are passing:
1. `RUN_SLOW=1 py.test tests/models/whisper/ -k speculative`
2. `py.test tests/ -k test_assisted_decoding_matches_greedy_search`
3. `py.test tests/ -k test_assisted_decoding_sample`

**Comment by jiqing-feng** [comment]
> Perfect, thank you for working on the changes 💪
> 
> @jiqing-feng if possible, it would be nice to delete the now unused `_extend_attention_mask` and `_extend_token_type_ids` functions :)
> 
> @amyeroberts I've confirmed on my end that all relevant tests are passing:
> 
> 1. `RUN_SLOW=1 py.test tests/models/whisper/ -k speculative`
> 2. `py.test tests/ -k test_assisted_decoding_matches_greedy_search`
> 3. `py.test tests/ -k test_assisted_decoding_sample`

Do you mean delete these 2 functions and replace all `_extend_xxx` functions with our new `_prepare_xxx` functions?

**Comment by gante** [comment]
@jiqing-feng yes `_extend_attention_mask` and `_extend_token_type_ids` -- are not used anywhere in the code

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating! 

@gante Thanks for running the tests! Do these also cover the tests that were breaking previously for whisper? Happy to merge once we know it's whisper compatible 🤗

**Comment by gante** [comment]
@jiqing-feng thank you for bearing with us 🤗

**Comment by amyeroberts** [comment]
@gante D'oh sorry - PR blindness 🤦  Thanks for merging and thanks again @jiqing-feng for all the work iterating on this PR!

**Pull request merged**
